### ISSUE TITLE
ORTModelForCausalLM batch inference is not consistent with infernece with batch size of 1

### ISSUE THREAD

**Issue by snehith-r** [huggingface/optimum#775]
I've a finetuned model of gpt2 to perform automated named entity recognition. For improving the inference speed I've used optimum to convert the model to onnx and use CUDAEXECUTION provide with IO bindings. With a batch size of 1 the attribute-value pairs are being extracted for each input, but If I use a batch size of 32 only attributes of a sequence are being generated. When I use batch size of 1 with the same 32 sequences it's extracting attribute-value pairs for each of the sequences. Idk what's causing failure in the inference of batch data. Any help would be appreciated.

TIA

**Comment by snehith-r** [comment]
It was due to the pad token argument in the generate function. But I've observed that onnx converted model is faster than pytorch with accelerate only if the batch size is 1 and pytorch with accelerate is outperforming the onnx inference on cuda if batch size is 32. Will share the code to reproduce once I check if it's the same with the pretrained Gpt2 provided in transformers library, for the scenario I mentioned it's a finetuned version trained on personal dataset.

**Comment by snehith-r** [comment]
@fxmarty  The code for comparing inference time, with increase in batch size(>1) onnx model is slower than pytorch
`from transformers import AutoTokenizer, GPT2TokenizerFast
from transformers import GPT2LMHeadModel,AdamW , GPT2Tokenizer, OpenAIGPTConfig

from transformers import AutoTokenizer

from optimum.onnxruntime import ORTModelForCausalLM
import torch

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

tokenizer = AutoTokenizer.from_pretrained("gpt2-large")
tokenizer.padding_side = "left"
tokenizer.pad_token = '<pad>'

sentneces=[ 'The people of the State of California do enact as follows:\n\n\nSECTION 1.\nSection 10295.35 is added to the Public Contract Code, to read:\n10295.35.\n(a) (1)', 
'Notwithstanding any other law, a state agency shall not enter into any contract for the acquisition of goods or services in the amount of one hundred thousand dollars ($100,000) or more with a contractor that, in the provision of benefits, discriminates between employees on the basis of an employees or dependent’s actual or perceived gender identity, including, but not limited to,' ,
'the employees or dependents identification as transgender.\n(2) For purposes of this section, “contract” includes contracts with a cumulative amount of one hundred thousand dollars ($100,000) or more per contractor in each fiscal year.\n(3) For purposes of this section, an employee health plan is discriminatory if the plan is not consistent with Section 1365.5 of the Health and Safety Code and Section 10140 of the Insurance Code.\n(4) The requirements of this section shall apply only to those portions of a contractors operations that occur under any of the following', 
'conditions:\n(A) Within the state.\n(B) On real property outside the state if the property is owned by the state or if the state has a right to occupy the property, and if the contractors presence at that location is connected to a contract with the state.\n(C) Elsewhere in the United States where work related to a state contract is being performed.\n(b) Contractors shall treat as confidential, to the maximum extent allowed by law or by the requirement of the contractors insurance provider, any request by an employee or applicant for employment benefits or any documentation of eligibility for benefits submitted by an employee or applicant for employment.\n(c) After taking all reasonable measures to find a contractor that complies with this section, as determined by the state agency, the requirements of this section may be waived under any of the following circumstances:\n(1)', 
'There is only one prospective contractor willing to enter into a specific contract with the state agency.\n(2) The contract is necessary to respond to an emergency, as determined by the state agency, that endangers the public health, welfare, or safety, or the contract is necessary for the provision of essential services, and no entity that complies with the requirements of this section capable of responding to the emergency is immediately available.\n(3) The requirements of this section violate, or are inconsistent with, the terms or conditions of a grant, subvention, or agreement, if the agency has made a good faith attempt to change the terms or conditions of any grant, subvention, or agreement to authorize application of this section.\n(4) The contractor is providing wholesale or bulk water, power, or natural gas, the conveyance or transmission of the same, or ancillary services,', 
'as required for ensuring reliable services in accordance with good utility practice, if the purchase of the same cannot practically be accomplished through the standard competitive bidding procedures and the contractor is not providing direct retail services to end users.\n(d) (1) A contractor shall not be deemed to discriminate in the provision of benefits if the contractor, in providing the benefits, pays the actual costs incurred in obtaining the benefit.\n(2) If a contractor is unable to provide a certain benefit, despite taking reasonable measures to do so, the contractor shall not be deemed to discriminate in the provision of benefits.\n(e) (1) Every contract subject to this chapter shall contain a statement by which the contractor certifies that the contractor is in compliance with this section.\n(2) The department or other contracting agency shall enforce this section pursuant', 
'to its existing enforcement powers.\n(3) (A) If a contractor falsely certifies that it is in compliance with this section, the contract with that contractor shall be subject to Article 9 (commencing with Section 10420), unless, within a time period specified by the department or other contracting agency, the contractor provides to the department or agency proof that it has complied, or is in the process of complying, with this section.\n(B) The application of the remedies or penalties contained in Article 9 (commencing with Section 10420) to a contract subject to this chapter shall not preclude the application of any existing remedies otherwise available to the department or other contracting agency under its existing enforcement powers.\n(f) Nothing in this section is intended to regulate the contracting practices of any local jurisdiction.\n(g) This section shall be construed so as not to conflict with applicable federal laws, rules, or regulations. In the event that a court or agency of competent jurisdiction holds that federal law, rule, or regulation invalidates any clause, sentence, paragraph, or section of this code or the application thereof to any person or circumstances, it is the intent of the state that the court or agency sever that clause, sentence, paragraph, or section so that the remainder of this section shall remain in effect.\nSEC. 2.\nSection 10295.35 of the Public Contract Code shall not be construed to create any new enforcement authority or responsibility in the Department of General Services or any other contracting agency.\nSEC. 3.\nNo reimbursement is required by this act pursuant to Section 6 of Article XIII\u2009B of the California Constitution because the only costs that may be incurred by a local agency or school district will be incurred because this act creates a new crime or infraction, eliminates a crime or infraction, or changes the penalty for a crime or infraction, within the meaning of Section 17556 of the Government Code, or changes the definition of a crime within the meaning of Section 6 of Article XIII\u2009B of the California Constitution.'
]
input_ids = tokenizer(sentneces, return_tensors="pt",padding=True).input_ids.to(device)
print(input_ids.shape)

model = ORTModelForCausalLM.from_pretrained("./gpt2LargeTitles",use_cache=False,from_transformers=True,provider= "CUDAExecutionProvider",use_io_binding=True).to(device)

# # save onnx checkpoint and tokenizer
onnx_path = "OpenAIGpt2"
model.save_pretrained(save_directory=onnx_path,decoder_file_name="model.onnx")
# tokenizer.save_pretrained(onnx_path)



import time
st = time.process_time()
gen_tokens = model.generate(input_ids,do_sample=True,temperature=0.5,max_length=input_ids.shape[1]+40,top_p=0.9,pad_token_id = tokenizer.encode(tokenizer.pad_token)[0],
                                        early_stopping=True)
et = time.process_time()
res = et - st
print('ONNX Generate Function Execution time:', res, 'seconds')
tokenizer.batch_decode(gen_tokens)

import time
# tokenizer = GPT2Tokenizer.from_pretrained('gpt2-large')
model_pytorch = GPT2LMHeadModel.from_pretrained('gpt2-large').to(device)
st = time.process_time()
gen_tokens = model_pytorch.generate(input_ids,max_length=input_ids.shape[1]+40,
                temperature=0.5,num_return_sequences=1, do_sample = True,
                                        top_p=0.9,
                                        pad_token_id = tokenizer.encode(tokenizer.pad_token)[0],
                                        early_stopping=True)
et = time.process_time()
res = et - st
print('Pytorch Generate Function Execution time:', res, 'seconds')
print(tokenizer.batch_decode(gen_tokens))`

**Comment by claeyzre** [comment]
Going from `OrtModelForSeq2SeqLM` to `AutoModelForSeq2SeqLM` fix this issue so I think it's an Optimum issue, maybe coming from the fact that attention_mask from the encoder is not propagated to the decoder which will take into account padded hidden states.

**Comment by claeyzre** [comment]
@michaelbenayoun maybe ?

**Comment by fxmarty** [comment]
@claeyzre Can confirm that the issue comes from not passing the encoder attention mask to the decoder, hence generating a meaningless mask automatically: https://github.com/huggingface/transformers/blob/v4.26.0/src/transformers/models/t5/modeling_t5.py#L965-L969

Should be fixed in https://github.com/huggingface/optimum/pull/1076
